### ISSUE TITLE
High level language feature ideas

### ISSUE THREAD

**Issue by BearishSun** [shader-slang/slang#535]
Hi, really like your project, it's about time we had a proper high-level shading language :smile:

I am developing a gamedev framework (https://github.com/GameFoundry/bsf) for which I have a custom shading language (BSL). It's not the greatest language as I don't have enough time to fully devote myself towards its development. But it does implement some features that I have found incredibly useful, and feel that are essential for a high level shading language. The features were all derived from actual use cases I had when designing bs::f's physically based renderer. So I'd thought I'd share some of them in case they are something you feel like integrating in Slang.

I'm not sure how much of this you might have already implemented, or planned, as I've just discovered the project, but here goes:
 - **Ability to override** - In bs::framework I have a renderer that has tiled deferred, clustered forward and standard deferred paths. I want to allow the user change the BRDF model, or even the lighting model (or even something else), but without them having to know anything about those three rendering paths. This saves the user from having to dive into some pretty complicated and specific shader code, having to write multiple shaders for multiple paths, and also ensures the renderer code can change later as long as the original interface didn't change.

   The way I handle it currently is that I have all shader code split into modules (I call them mixins), which can then be included in a shader. But if a mixin with the same name is declared later in a file, it will override the previous one, allowing users to just include a file and then override a mixin they desire (e.g. "BRDF" or "SpotLightAttenuation" mixin). In BSL I don't have interfaces so there's no nice way to check if the user is conforming to what the mixin expects, which is a downside of the current approach. With support for interfaces I can see Slang implementing this much more neatly.

 - **Variations** - Shaders very often need to be compiled using multiple different permutations. One example is having a vertex shader that can process static, bone animated, morph animated or bone + morph animated vertices. Or some effect with different quality levels. Or a fragment shader with or without MSAA support. All of those permutations normally share 75%+ of the code, the rest usually surrounded with preprocessor conditionals.

   In most frameworks/engines this is handled though #define's which are set from non-shader code and trigger shader compilation for all possible permutations. I found there is significantly less non-shader boilerplate code if you set the variations in the shader itself. Especially when there are a lot of variations the number of permutations goes up quickly, and since you need to manually set all #define's externally, you need to do it for each variation, usually resulting in use of ugly templates and macros (at best).

   The compiler could instead pick up all the possible variations and output different versions of the shader, with some meta-information identifying the variation. Using the traditional #define approach you also need to recompile the entire shader for each permutation, but with this information available to the compiler it could potentially re-use the AST and output all variations at once? (That would be an optional bonus feature, rather than a requirement).

   You can also #include or inherit variations from other modules (mixins). In the example with the vertex shader above, if user wanted to modify the vertex position (e.g. make it wavy via sin(time)), he could just include the "VertexInput" mixin, which already defines the four relevant variations. He only needs to implement the interface without needing to know this shader actually has four different vertex processing paths, nor does he need to let non-shader code know this shader processes vertices and needs to be compiled with special #defines.

   In BSL I'm doing this in a very simple way. First pass just looks for variations, which look like this:
   ```
   variations
   {
   	MSAA_COUNT = { 1, 2 };
   	QUALITY = { 0, 1, 2, 3, 4 };
   	MSAA_RESOLVE_0TH = { true, false };
   };	
   ```

   Then all permutations are generated and the shader is fully re-parsed for each variation, while setting the relevant #define. Variation specific code then just uses normal preprocessor conditionals. A non-preprocessor syntax might be preferable - I chose it just because it was simpler to implement.

 - **Non-programmable states** - Ability to set non-programmable states like blend, rasterizer, depth-stencil in the shader code itself. Similar to HLSL effects framework, but ideally better syntax. It's much nicer to have these defined with the rest of the shader, rather than in non-shader code. 
 - **Default values** - Ability to specify default values for primitive types, as well as textures and samplers. Textures can support a pre-defined set of values like "black", "white", "normalYUp" or similar. This makes it easier for user-created materials to "just work", even if the artist fails to set all textures (e.g. he sets just the albedo texture, but no normal/metalness/roughness). These default values would just be reported through reflection, but otherwise not output to HLSL/GLSL.
 
If some these are not clear, I do have a short guide on BSL here: https://github.com/GameFoundry/bsf/blob/master/Documentation/Manuals/User/bsl.md

You can also find how the mixin override system works in practice here:
https://github.com/GameFoundry/bsf/blob/master/Documentation/Manuals/User/surfaceShaders.md

Like I said, I really like this project, and I'm not fully happy with my own language. I just wanted to suggest some ideas of features that I have found very useful and perhaps you can integrate them in the future? 

I'd certainly love to switch to Slang for my framework in the future :sunglasses:

**Comment by tfoleyNV** [comment]
High-order bits:

* Many of the things you describe are things we would like to support some day in Slang, although the exact way we expose things might differ from how BSL does it.

* We have a lot of work still to do on our core feature set, so our priority right now is on making the existing language features and cross-compilation more robust before we start adding more complexity to the existing language

More details:

For a bit of context, the Slang system is actually derived from our experience working on some research shading languages that had more of the features you describe. For example, [my graduate work](https://graphics.stanford.edu/papers/spark/) in this area was a shading langauge that had overrides/inheritance and support for driving non-programmable states.

One constraint we've placed on the Slang effort so far is that our language should avoid inventing new syntax or new language constructs whenever possible, and should instead make use of stuff from existing *general-purpose* languages instead. The goal there is to reduce the barriers to adoption for new users, since they aren't having to learn a language with all kinds of special-purpose syntax/semantics.

That doesn't mean we can't support things like your `shader` declarations, but it might be done with existing language constructs instead, e.g.:

```hlsl
// over-simplified example of a shader "pipeline" implemented as a struct
struct MyPipeline: IRasterizationPipeline
{
    // uniform shader parameters are expressed as members
    Texture2D diffuseMap;
    ParameterBlock<Camera> camera;
    // ...

    // cross-stage data gets declared as nested structs
    // (actually associated types of the IRasterizationPipeline interface)
    struct V2F { float3 position; float3 normal };

    // per-stage entry points are methods
    // (which satisfy requirements of the interface)
    V2F vertexStage()
    { ... }
    float4 fragmentStage()
    { ... }
}
```

In that example, we've used an ordinary `struct` to express something similar in spirit (if not in all the details) to a BSL `shader` declaration (we've collapsed the `shader` and its nested `code`...).

With that bit of preliminary out of the way, I'll try to run down your list of feature proposals:

#### Ability to Override

We have considered allowing `override` declarations in types, for cases where they inherit from a type that provides default implementations. We are a bit wary of adding full inheritance for `struct` types (one `struct` inheriting from another), and we aren't ready to add `class` types right now (because they should ideally have reference semantics). The two places where inheritance and override could come into the picture, then, are:

1. If we start allowing `interface`s to define default implementations of their requirements, then a type that implements the interface would be able to override those defaults.

2. We have previously discussed adding a `mixin` construct similar to what you describe in BSL, at which point we'd use a linearization step to determine a full "method resolution order" for a `struct` + all the mixins it inherits from.

It sounds like the examples you've built in BSL would require option (2) there to work. We'd like to get there at some point, but don't go holding your breath. :)

#### Variations

Our current assumption is that axes of variation should be expressed using generics in Slang (whether generic type parameters or generic value parameters), and that variants of a shader entry point are then constructed by tools.

In your case, it seems like you'd want a way to decorate generic value parameters with the ranges/values they might take on, and use this to instruct either the Slang compiler or a layered renderer-specific tool to generate versions specialized to those parameters.

This is within the scope of what we'd like to support, just with the change where we'd prefer generic value parameters over `#define`s, because they can work with separate compilation.

#### Non-Programmable States

This is definitely a nice-to-have feature, and I've implemented stuff like this before in the system I linked above. Adding this to Slang some day is in the realm of possibility, but there are a few challenges:

* The constraints on what can be expressed in fixed-function state end up being API- or renderer-specific, so Slang could end up in the business of defining a one-size-fits-all state model and then mapping it to various target APIs. That seems to be beyond what the project should encompass (e.g., a nice line in the sand today is that the Slang compiler implementation doesn't include the headers for any graphics APIs).

* The interface for how to expose that information to an engine could get complex. In the simple case it might just be a set of key/value pairs ("what was the setting, if any, for state X?"), but powerful engines might like to support cases where some pieces of fixed-function state are set as a general expression of engine/application state (e.g., `cullBackfaces = engineModel->material->isDoubleSided`).

Such an interface would need to be evolved carefully with input from at least a few renderers, to make sure that Slang wouldn't bind too much to what a single renderer needs.

#### Default Values

As I understand it, you are mostly concerned with default values on global shader parameters and `cbuffer` members. We'd like to support those at least in the easy cases (e.g., `float4 albedo = float4(1.0f);` should ideally work and be exposed through reflection).

Once you want to expose default values for a `Texture2D` parameter, you run into similar problems as for non-programmable state: the exact capabilities and requirements might be engine-specific. Some engine is going to want to be able to do `Texture2D gPrevFrame = renderer.prevFrameBuffers.colorTarget.srv;` or something like that.

In the limit, the challenge is that you don't just want an API for exposing a *value* for a default parameter, but rather an *expression* for computing the default value (even your example of a simple `white` texture is really a symbolic expression that the shader compiler must pass along to the engine). The question then is: what form should the reflection API expose for such expressions? A parsed AST (behind an API)? The generated IR code for the initializer expression (+ a small VM for executing it)?

---

Long story short, the stuff you describe from BSL is all reasonable and we definitely want some similar capabilities in Slang. The biggest reason we don't have this stuff today is time (there isn't enough of it), and the second biggest reason is that figuring out how to make these features fit in cleanly without special-case syntax/semantics is really hard work (and usually makes the implementation that much harder).

If you know of anybody who would be interested in implementing some of these features in the context of Slang, we'd be happy to coordinate with them on the design and planning.

**Comment by BearishSun** [comment]
Thanks for the detailed response. I'm glad you want similar capabilities in Slang, it certainly makes me even more excited about the project :)

I don't have any answers for the design issues you outlined. Personally even if you just spit out a part of the AST I need to process myself, I'd be happy :) Since some of these features feel like they will be hard to do in a really general purpose way, perhaps consider implementing them as extensions? You could relax the your general-purpose design requirement since it wouldn't be a part of the core language. Anything engine-specific could probably be done through some kind of general purpose API for dealing with the parsed AST, as you mentioned.

BTW, I really like the example of being able to hook up engine specific state in the shader itself, it's not something I have considered, but it certainly feels like it could be very useful. 

If I wasn't busy myself I'd certainly help out with the project. I definitely had plans on reworking a lot of BSL internals, so when that time comes I might come here and work on Slang instead - but that might be a while considering the amount of work I have ahead!
